hermes-frontend/src/test/groovy/pl/allegro/tech/hermes/frontend/producer/kafka/KafkaBrokerMessageProducerIntegrationTest.groovy

```diff
@@ -0,0 +1,205 @@
+import static org.apache.kafka.clients.producer.ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG
+
+@Testcontainers
+class KafkaBrokerMessageProducerIntegrationTest extends Specification {
+
+    static Integer NUMBER_OF_PARTITION = 3
+
+    @Shared
+    KafkaContainer kafkaContainer = new KafkaContainer()
+
+    @Shared
+    ConfigFactory configFactory = Mock() {
+        getLongProperty(Configs.FRONTEND_STARTUP_WAIT_KAFKA_INTERVAL) >> 1L
+        getStringProperty(Configs.KAFKA_HEADER_NAME_MESSAGE_ID) >> "id"
+        getStringProperty(Configs.KAFKA_HEADER_NAME_TIMESTAMP) >> "ts"
+        getStringProperty(Configs.KAFKA_HEADER_NAME_SCHEMA_VERSION) >> "sv"
+        getStringProperty(Configs.KAFKA_HEADER_NAME_SCHEMA_ID) >> "sid"
+    }
+
+    @Shared
+    KafkaProducer<byte[], byte[]> leaderConfirms
+
+    @Shared
+    KafkaProducer<byte[], byte[]> everyoneConfirms
+
+    @Shared
+    KafkaBrokerMessageProducer brokerMessageProducer
+
+    @Shared
+    KafkaNamesMapper kafkaNamesMapper = new JsonToAvroMigrationKafkaNamesMapper("")
+
+    @Shared
+    Producers producers
+
+    @Shared
+    String containerId
+
+    @Shared
+    AdminClient adminClient
+
+
+    def setupSpec() {
+        kafkaContainer.start()
+        kafkaContainer.waitingFor(Wait.forHealthcheck())
+        containerId = kafkaContainer.containerId
+
+        Properties props = new Properties()
+        props.put(BOOTSTRAP_SERVERS_CONFIG, kafkaContainer.bootstrapServers)
+        props.put(KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
+        props.put(VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
+        leaderConfirms = new KafkaProducer<byte[], byte[]>(props)
+        everyoneConfirms = new KafkaProducer<byte[], byte[]>(props)
+        adminClient = AdminClient.create(props)
+    }
+
+    def setup() {
+        producers = new Producers(leaderConfirms, everyoneConfirms, configFactory)
+        brokerMessageProducer = new KafkaBrokerMessageProducer(producers,
+                new HermesMetrics(new MetricRegistry(), new PathsCompiler("localhost")),
+                new KafkaHeaderFactory(configFactory),
+                configFactory)
+    }
+
+    def "should publish messages on only one partition for the same partition-key"() {
+        Topic topic = createAvroTopic("pl.allegro.test.Foo")
+        Subscription subscription = createTestSubscription(topic, "test-subscription")
+        String kafkaTopicName = topic.getName().toString()
+        ConsumerGroupId consumerGroupId = kafkaNamesMapper.toConsumerGroupId(subscription.qualifiedName)
+        createTopicInKafka(kafkaTopicName, NUMBER_OF_PARTITION)
+        CachedTopic cachedTopic = CachedTopicsTestHelper.cachedTopic(topic)
+
+        when:
+        1.upto(10) {
+            brokerMessageProducer.send(generateAvroMessage("partition-key"), cachedTopic, null)
```

Comment:
You send messages asynchronously here and assume that they are already in Kafka, when `then` section is executed. At first glance, this test can be unstable, especially in unstable environment such as our CI. To stabilize it, I suggest to wait for messages being  published successfully on Kafka and then check offsets on the Kafka topic. This comment also applies to the second test as well.

Reply:
I've modified tests by adding a waiting mechanism. Please let me know whether it's ok for you :)